File: educational-core/src/com/jetbrains/edu/coursecreator/projectView/CCTaskNode.java
package com.jetbrains.edu.coursecreator.projectView;

import com.intellij.ide.projectView.ViewSettings;
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode;
import com.intellij.ide.util.treeView.AbstractTreeNode;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiNamedElement;
import com.jetbrains.edu.learning.*;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.ext.TaskExt;
import com.jetbrains.edu.learning.courseFormat.tasks.Task;
import com.jetbrains.edu.learning.projectView.TaskNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CCTaskNode extends TaskNode {
  public CCTaskNode(@NotNull Project project,
                    PsiDirectory value,
                    ViewSettings viewSettings,
                    @NotNull Task task) {
    super(project, value, viewSettings, task);
  }

  @Nullable
  @Override
  public AbstractTreeNode modifyChildNode(AbstractTreeNode childNode) {
    AbstractTreeNode node = super.modifyChildNode(childNode);
    if (node != null) {
      return node;
    }
    Object value = childNode.getValue();
    if (value instanceof PsiDirectory) {
      String name = ((PsiDirectory) value).getName();
      if (EduNames.BUILD.equals(name) || EduNames.OUT.equals(name)) {
        return null;
      }

      if (name.equals(TaskExt.getSourceDir(myTask)) || name.equals(TaskExt.getTestDir(myTask))) {
        return createChildDirectoryNode((PsiDirectory) value);
      }
    }
    if (value instanceof PsiElement) {
      PsiElement psiElement = (PsiElement) value;
      PsiFile psiFile = psiElement.getContainingFile();
      VirtualFile virtualFile = psiFile.getVirtualFile();
      if (virtualFile == null) {
        return null;
      }
      Course course = StudyTaskManager.getInstance(myProject).getCourse();
      if (course == null) {
        return null;
      }
      EduConfigurator configurator = EduConfiguratorManager.forLanguage(course.getLanguageById());
      if (configurator == null) {
        return new CCStudentInvisibleFileNode(myProject, psiFile, getSettings());
      }
      if (!EduUtils.isTestsFile(myProject, virtualFile)) {
        return new CCStudentInvisibleFileNode(myProject, psiFile, getSettings());
      }
      return new CCStudentInvisibleFileNode(myProject, psiFile, getSettings(), getTestNodeName(configurator, psiElement));
    }
    return null;
  }

  @NotNull
  private static String getTestNodeName(EduConfigurator configurator, PsiElement psiElement) {
    String defaultTestName = configurator.getTestFileName();
    if (psiElement instanceof PsiFile) {
      return defaultTestName;
    }
    if (psiElement instanceof PsiNamedElement) {
      String name = ((PsiNamedElement)psiElement).getName();
      return name != null ? name : defaultTestName;
    }
    return defaultTestName;
  }

  @Override
  public PsiDirectoryNode createChildDirectoryNode(PsiDirectory value) {
    return new CCNode(myProject, value, getSettings());
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/projectView/CCTaskNode.kt
package com.jetbrains.edu.coursecreator.projectView

import com.intellij.ide.projectView.ViewSettings
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode
import com.intellij.ide.util.treeView.AbstractTreeNode
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiNamedElement
import com.jetbrains.edu.coursecreator.courseignore.CourseIgnoreRules
import com.jetbrains.edu.learning.EduNames
import com.jetbrains.edu.learning.StudyTaskManager
import com.jetbrains.edu.learning.configuration.EduConfigurator
import com.jetbrains.edu.learning.courseFormat.ext.configurator
import com.jetbrains.edu.learning.courseFormat.tasks.Task
import com.jetbrains.edu.learning.isTestsFile
import com.jetbrains.edu.learning.projectView.TaskNode

class CCTaskNode(
  project: Project,
  value: PsiDirectory,
  viewSettings: ViewSettings,
  private val courseIgnoreRules: CourseIgnoreRules,
  task: Task
) : TaskNode(project, value, viewSettings, task) {

  override fun modifyChildNode(childNode: AbstractTreeNode<*>): AbstractTreeNode<*>? {
    val node = super.modifyChildNode(childNode)
    if (node != null) {
      return node
    }
    val value = childNode.value
    if (value is PsiDirectory) {
      val name = value.name
      if (EduNames.BUILD == name || EduNames.OUT == name) return null
      return createChildDirectoryNode(value)
    }
    else if (value is PsiElement) {
      val psiFile = value.containingFile
      val virtualFile = psiFile.virtualFile ?: return null
      val course = StudyTaskManager.getInstance(myProject).course ?: return null
      val configurator = course.configurator ?: return CCStudentInvisibleFileNode(myProject, psiFile, settings, courseIgnoreRules)
      return if (!virtualFile.isTestsFile(myProject)) {
        CCStudentInvisibleFileNode(myProject, psiFile, settings, courseIgnoreRules)
      }
      else {
        CCStudentInvisibleFileNode(myProject, psiFile, settings, courseIgnoreRules, getTestNodeName(configurator, value))
      }
    }
    return null
  }

  override fun createChildDirectoryNode(value: PsiDirectory): PsiDirectoryNode {
    return CCNode(myProject, value, settings, courseIgnoreRules, item)
  }

  companion object {
    private fun getTestNodeName(configurator: EduConfigurator<*>, psiElement: PsiElement): String {
      val defaultTestName = configurator.testFileName
      if (psiElement is PsiFile) {
        return defaultTestName
      }
      if (psiElement is PsiNamedElement) {
        val name = psiElement.name
        return name ?: defaultTestName
      }
      return defaultTestName
    }
  }
}
